### PR TITLE
Feature/auto exposure

### PR DESCRIPTION
**Pull request by supitalp**
This PR implements a simple auto-exposure for the frame-based sensor. This is a simple proportional controller that tries to ensure that the (alpha-trimmed) mean intensity is 128.

### CODE REVIEW

**Comment by supitalp** [comment]
@uzh-rpg/toni-rosinol ptal

File: davis_ros_driver/src/driver.cpp [hunk 1]
```diff
@@ -118,521 +120,565 @@ void DavisRosDriver::dataStop()
...
+                    const int numEvents = caerEventPacketHeaderGetEventNumber(packetHeader);
+                    for (int j = 0; j < numEvents; j++)
+                    {
+                        // Get full timestamp and addresses of first event.
+                        caerPolarityEvent event = caerPolarityEventPacketGetEvent(polarity, j);
+
+                        dvs_msgs::Event e;
+                        e.x = caerPolarityEventGetX(event);
+                        e.y = caerPolarityEventGetY(event);
+                        e.ts = reset_time_ + ros::Duration(caerPolarityEventGetTimestamp64(event, polarity) / 1.e6);
```
Comment (ToniRV) [inline davis_ros_driver/src/driver.cpp]: Hey @supitalp, I remember that Timo actually did a commit changing the way timestamps are computed. It is this [one](https://github.com/uzh-rpg/rpg_dvs_ros/commit/6474994c5736d46213096c2164bc0eee745eb384). I would personally do it as he did, I also experienced weird floating behavior when doing this instead. Basically, 1e6 has type double whether 1000000 has type int.

Comment (ToniRV) [inline davis_ros_driver/src/driver.cpp]: Now I see that this was here before, I would use this opportunity to change it though

File: davis_ros_driver/src/driver.cpp [hunk 2]
```diff
@@ -118,521 +120,565 @@ void DavisRosDriver::dataStop()
...
+                            msg.angular_velocity.x = -caerIMU6EventGetGyroX(event) / 180.0 * M_PI;
+                            msg.angular_velocity.y = caerIMU6EventGetGyroY(event) / 180.0 * M_PI;
+                            msg.angular_velocity.z = -caerIMU6EventGetGyroZ(event) / 180.0 * M_PI;
+                        }
+
+                        // no orientation estimate: http://docs.ros.org/api/sensor_msgs/html/msg/Imu.html
+                        msg.orientation_covariance[0] = -1.0;
+
+                        // time
+                        msg.header.stamp = reset_time_ + ros::Duration(caerIMU6EventGetTimestamp64(event, imu) / 1.e6);
```
Comment (ToniRV) [inline davis_ros_driver/src/driver.cpp]: Same here

File: davis_ros_driver/src/driver.cpp [hunk 3]
```diff
@@ -118,521 +120,565 @@ void DavisRosDriver::dataStop()
...
+                    {
+                        for (int img_x=0; img_x<frame_width; img_x++)
+                        {
+                            const uint16_t value = image[img_y*frame_width + img_x];
+                            msg.data.push_back(value >> 8);
+                        }
+                    }
+
+                    // time
+                    msg.header.stamp = reset_time_ + ros::Duration(caerFrameEventGetTimestamp64(event, frame) / 1.e6);
```
Comment (ToniRV) [inline davis_ros_driver/src/driver.cpp]: Same here.

File: davis_ros_driver/src/driver.cpp [hunk 4]
```diff
@@ -118,521 +120,565 @@ void DavisRosDriver::dataStop()
...
+}
 
-          // time
-          msg.header.stamp = reset_time_ + ros::Duration(caerFrameEventGetTimestamp64(event, frame) / 1.e6);
+int DavisRosDriver::computeNewExposure(const std::vector<uint8_t>& img_data, const uint32_t current_exposure) const
+{
+    constexpr float desired_intensity = 128;
+    constexpr int min_exposure = 10;
+    constexpr int max_exposure = 150000;
+    constexpr float proportion_to_cut = 0.25f;
```
Comment (ToniRV) [inline davis_ros_driver/src/driver.cpp]: I would make the above values static as they are const at compile time.

**Comment by ToniRV** [comment]
LGTM, some comments on timestamps to address and a conflict to resolve, but besides this :+1:

File: davis_ros_driver/src/driver.cpp (hunk 1 above)
Comment (supitalp) [inline davis_ros_driver/src/driver.cpp]: OK, I will do that at the same time as I resolve the conflict with the commit you mention.

File: davis_ros_driver/src/driver.cpp (hunk 2 above)
Comment (supitalp) [inline davis_ros_driver/src/driver.cpp]: ok

File: davis_ros_driver/src/driver.cpp (hunk 3 above)
Comment (supitalp) [inline davis_ros_driver/src/driver.cpp]: ok

File: davis_ros_driver/src/driver.cpp (hunk 4 above)
Comment (supitalp) [inline davis_ros_driver/src/driver.cpp]: ok, thanks

**Comment by supitalp** [comment]
@ToniRV did the changes you asked for and fixed the conflicts with the master branch. could you please approve the PR?

**Pull request merged**